riscv-page-tables/src/page_table.rs

```diff
@@ -1046,6 +1053,66 @@ impl<T: PagingMode> GuestStagePageTable<T> {
         self.do_invalidate_range(vaddr, len, true, pred)
     }
 
+    fn do_validate_range<F>(
+        &self,
+        vaddr: PageAddr<T::MappedAddressSpace>,
+        len: u64,
+        mut pred: F,
+    ) -> Result<impl Iterator<Item = SupervisorPageAddr> + '_>
+    where
+        F: FnMut(SupervisorPageAddr) -> bool,
+    {
+        let num_pages = PageSize::num_4k_pages(len);
+        vaddr
+            .checked_add_pages(num_pages)
+            .ok_or(Error::AddressOverflow)?;
+        let mut inner = self.inner.lock();
+        for va in vaddr.iter_from().take(num_pages as usize) {
+            use TableEntryType::*;
+            match inner.walk(va.into()) {
+                Invalidated(pte) => {
+                    if !pte.level().is_leaf() {
+                        return Err(Error::PageSizeNotSupported(pte.level().leaf_page_size()));
+                    }
+                    if !pred(pte.page_addr()) {
+                        return Err(Error::PredicateFailed);
+                    }
+                }
+                Unused(_) => continue,
```

Comment:
Oh so do you also expect that we should be blocking a fully populated range? I'm asking because then I would use the non-sparse version of `invalidate_range()`.